File: java/diamond/src/main/java/DiamondPrinter.java
import java.util.ArrayList;
import java.util.List;

class DiamondPrinter {

    List<String> printToList(char a) {
        final var side = a - 'A' + 1;
        final var rows = side * 2 - 1;

        final var result = new ArrayList<String>();

        for (var r = 0; r < rows; r++) {
            final var d = Math.min(r, rows - 1 - r);
            final var c = (char) (d + 'A');

            final var outerSpaces = side - d - 1;

            final var sb = new StringBuilder();

            sb.repeat(' ', outerSpaces);
            sb.append(c);

            if (d > 0) {
                sb.repeat(' ', 2 * d - 1);
                sb.append(c);
            }

            sb.repeat(' ', outerSpaces);
            result.add(sb.toString());
        }

        return result;
    }

}

File: kotlin/diamond/src/main/kotlin/DiamondPrinter.kt
import kotlin.math.abs

class DiamondPrinter {

    fun printToList(size: Char): List<String> {
        val side = size - 'A' + 1

        val rows = side * 2 - 1
        val columns = side * 2 - 1

        return buildList {
            repeat(rows) { i ->
                val d = minOf(i, rows - 1 - i)
                val c = (d + 'A'.code).toChar()

                add(
                    buildString {
                        repeat(columns) { j ->
                            when (abs(j - columns / 2)) {
                                d -> append(c)
                                else -> append(" ")
                            }
                        }
                    }
                )
            }
        }
    }
}
